cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/strategy/AbstractReplicaMovementStrategy.java

```diff
@@ -0,0 +1,84 @@
+ */
+
+package com.linkedin.kafka.cruisecontrol.executor.strategy;
+
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionProposal;
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionTask;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import org.apache.kafka.common.Cluster;
+
+
+/**
+ * An abstract class for replica movement strategy. This class will be extended to create custom strategy to determine the
+ * execution order the replica movement tasks.
+ */
+public abstract class AbstractReplicaMovementStrategy implements ReplicaMovementStrategy {
+
+  /**
+   * Generate a comparator for replica movement task which incorporate the strategy to apply. The "smaller" task will have
+   * higher execution priority.
+   *
+   * @param cluster The current cluster state.
+   * @return The comparator of task.
+   */
+  public abstract Comparator<ExecutionTask> taskComparator(Cluster cluster);
+
+  @Override
+  public ReplicaMovementStrategy chain(ReplicaMovementStrategy strategy) {
+    if (!(strategy instanceof AbstractReplicaMovementStrategy)) {
+      throw new IllegalArgumentException("Replica movement strategy " + this.getClass().getSimpleName() + " is unable to chain with "
+                                         + strategy.getClass().getSimpleName() + ".");
+    }
+    AbstractReplicaMovementStrategy current = this;
+    return new AbstractReplicaMovementStrategy() {
+      @Override
+      public Comparator<ExecutionTask> taskComparator(Cluster cluster) {
+        Comparator<ExecutionTask> comparator1 = current.taskComparator(cluster);
+        Comparator<ExecutionTask> comparator2 = ((AbstractReplicaMovementStrategy) strategy).taskComparator(cluster);
+
+        return (task1, task2) -> {
+          int compareResult1 = comparator1.compare(task1, task2);
+          int compareResult2 = comparator2.compare(task1, task2);
+          if (compareResult1 == 0) {
+            return compareResult2;
+          }
+          return compareResult1;
+        };
+      }
+    };
+  }
+
+  @Override
+  public Map<Integer, SortedSet<ExecutionTask>> applyStrategy(Collection<ExecutionTask> replicaMovementTasks, Cluster cluster) {
+    Map<Integer, SortedSet<ExecutionTask>> tasksByBrokerId = new HashMap<>();
+
+    for (ExecutionTask task : replicaMovementTasks) {
+      ExecutionProposal proposal = task.proposal();
+
+      // Add the task to source broker's execution plan
+      int sourceBroker = proposal.oldLeader();
+      SortedSet<ExecutionTask> sourceBrokerTaskSet = tasksByBrokerId.computeIfAbsent(sourceBroker,
+                                                                                     k -> new TreeSet<>(taskComparator(cluster)));
+      if (!sourceBrokerTaskSet.add(task)) {
```

Comment:
If `replicaMovementTasks` is a `Set` (please see my previous comment), then this check would be redundant.

Reply:
This check is still needed because the way `SortedSet` works is it use passed-in `Comparator.compare(a, b)==0` instead of item's `equal()` to determine whether to replace an existing item. So it highly depends on the `Comparator` which `AbstractReplicaMovementStrategy` provides.

one example is 
  public void test2() {
    SortedSet<Integer> ss = new TreeSet<>((o1, o2) -> {
      return 0;
    });
    ss.add(1);
    assertFalse(ss.add(2));
  }

Reply:
Good point!
But, then what prevents any of our existing strategies, such as `PostponeUrpReplicaMovementStrategy#taskComparator`, to fail this sanity check unless `BaseReplicaMovementStrategy#taskComparator` is also chained in the strategies? Do we have an implicit assumption that the user always provides `BaseReplicaMovementStrategy` as one of the required strategies?

If that is the case, I think we should (1) indeed always use `BaseReplicaMovementStrategy` as the last chained strategy in all cases, and (2) not make it a configuration.

Reply:
My original idea does not have the assumption that `the user always provides BaseReplicaMovementStrategy as one of the required strategies`.  

The user is free to apply whatever rule they want and if they can guarantee that comparator they implement never return 0, they do not need to chain with `BaseReplicaMovementStrategy`(one example can be alphabetic order of `TopicPartition`). And if they miss the guarantee in their implementation, Cruise Control will not misbehave silently because we have this sanity check.

It is true the three strategies we provide , `PostponeUrpReplicaMovementStrategy`, `PrioritizeLargeReplicaMovementStrategy` and `PrioritizeSmallReplicaMovementStrategy` do not have this guarantee and I think it is ok because in the configuration we ask for a list of strategy. so if the client want to do things like `postpone urp replica movement and if two movement are urp, execute the one with smaller task id.` they should explicitly provide 2 strategies in the configure because it is indeed two rules here.